.travis.yml

```diff
@@ -1,17 +1,70 @@
 # we use both C and C++, so advertize C++
 language: cpp
 cache: ccache
-dist: trusty
-compiler:
-  - gcc
 env:
   - GTK3=no BINRELOC=no
   - GTK3=yes BINRELOC=no
   - GTK3=no BINRELOC=yes
   - GTK3=yes BINRELOC=yes
   - GTK3=no MINGW=yes
   - GTK3=yes MINGW=yes
+matrix:
+# gcc
+  include:
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-5
+      env:
+         - MATRIX_EVAL="CC=gcc-5 && CXX=g++-5"
+
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-6
+      env:
+        - MATRIX_EVAL="CC=gcc-6 && CXX=g++-6"
+
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-7
+      env:
+        - MATRIX_EVAL="CC=gcc-7 && CXX=g++-7"
+
+    - os: linux
+      dist: xenial
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-8
+      env:
+        - MATRIX_EVAL="CC=gcc-8 && CXX=g++-8"
+
+# clang        
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - llvm-toolchain-trusty-5.0
+          packages:
+            - clang-5.0
+      env:
+        - MATRIX_EVAL="CC=clang-5.0 && CXX=clang++-5.0"
+
 before_install:
+  - eval "${MATRIX_EVAL}"
```

Comment:
this should be in `before_script` below with the `CFLAGS` export, shouldn't it?

Reply:
Also I'd rather see this done using `export`, and set e.g. `MATRIX_EXPORT="CC=clang-5.0 CXX=clang++-5.0"`, which is effectively what we want done (and so it doesn't rely on `CC` and `CXX` being exported by something else).

Reply:
As @elextr noted, what I've used is suggested by the travis docs. It doesn't appear from the logs that CC and CXX are getting exported by anything else later. But you're concern is that it's a possibility? But definitely let me know if you're sure you'd like me to change that anyway.